[Diagnostics] Add educational notes explaining multiple inheritance (#38253)
apple/swift: userdocs/diagnostics/multiple-class-inheritance.md

> # Multiple Inheritance of Classes

Reviewer: I'd suggest either just "Multiple Inheritance" or "Inheritance from Multiple Classes" if you want to elaborate, because "of" here is kind of ambiguous.

> In some programming languages, a class can have multiple superclasses. This feature is called multiple inheritance and is unsupported in Swift. However, using protocols, one can achieve a similar effect as multiple inheritance.

Reviewer: ```suggestion
In some programming languages, a class can inherit the interface of multiple superclasses. Known as multiple inheritance, this feature can add significant complexity to the language and is unsupported in Swift. Instead, Swift allows composition of interfaces using protocols.
```

I think it's important to state (however briefly or generally) a rationale why something isn't supported. Readers are naturally inclined to seek out reasons. Without something there, the text raises ambiguity as to whether this lack of support is temporary and unintentional, or even if the reader correctly surmises that it is intentional, leaves them guessing as to why.

> 
> # Multiple Inheritance of Classes
> In some programming languages, object can inherit features and characteristics from multiple parent or base classes. This feature is called multiple inheritence and is unsupported in swift. However, Swift provide language features that can acheive results similar to multiple inheritence through the use of protocols.

Reviewer 2: ```suggestion
# Multiple Inheritance of Classes

In some programming languages, a class can have multiple superclasses. This feature is called multiple inheritance and is unsupported in Swift. However, using protocols, one can achieve a similar effect as multiple inheritance.
```

> Consider the following example with a class `Bird` conforming to two protocols `Animal` and `Flyable`.

Reviewer: ```suggestion
Consider the following example:
```

While this example is simple, I don't think that it's the most effective demonstration here of the concept of protocol composition. Here's why:

The classic `Animal` example is often used because there is a well-known taxonomy among species that doesn't require explanation, because the reader can be expected to know this from grade school. However, that taxonomy precisely doesn't allow the kind of composition for which you might want to use multiple inheritance: no animal is both an amphibian and a mammal, for example.

Moreover, using an `Animal` example could be misleading to the reader whose mind goes to taxonomy, since inheriting from both `Animal` and `Mammal` notably would _not_ be an example of multiple inheritance, because `Mammal` would inherit from `Animal`.

Finally, nothing in the text here demonstrates a rationale for why `Flyable` exists; I don't think this evokes any ideas in the reader like `Animal` does, so to justify the protocol you'd have to provide some rationale why you'd want to model `Flyable`s in code orthogonally to the species taxonomy--which is certainly doable, but the example becomes much wordier.

I'm sure there are many commonly used examples from C++ pedagogy, say, that could be drawn upon for inspiration here. I don't know of any, but one example I can think of from the top of my head would be:

```swift
protocol Utensil { var name: String }

protocol ServingUtensil: Utensil { func serve() }
protocol Spoon: Utensil { func scoop() }
protocol Fork: Utensil { func spear() }

struct Ladle: ServingUtensil, Spoon { ... }
struct CarvingFork: ServingUtensil, Fork { ... }
struct Spork: Spoon, Fork { ... }
```

Author: A reasonable suggestion, I will adapt your suggestions by using the utensil example.

```swift
protocol Animal {
    func move()
}

class Bird: Animal, Flyable {
    func move() {
        print("Moving...")
    }
}

> ```

Reviewer 2: Could you add the definition of `Flyable` to the code example? Without it, it's less clear what you're trying to show here.

Author: Done.

    func move()
}

extension Flyable {
    func fly() {
        print("Flying...")
    }
}

class Bird: Animal, Flyable {
    // move() and fly() use default implementations
}
```

> Since `Animal` and `Flyable` provide default implementations for their protocol requirements, `Bird` can conform to both without implementing `fly()` or `move()` and the default implementation will be used. This effect is similar to class inheritance where a subclass inherits properties, subscripts and methods from a superclass. The default implementation can be overriden by defining `fly()` or `move()` without an `override` keyword.

Reviewer: ```suggestion
```

As noted previously, I do not think this educational note needs to explain or demonstrate how default implementations work. It's pretty inherent to basic usage of protocols; if a user doesn't know about protocols, they can follow the links given below to read more in _TSPL_ which explain it adequately.

> There is one caveat here on the interaction with subclassing. If a superclass conforms a protocol but does not override the default implementation of a protocol requirement, its subclasses cannot override the default implementation ([SR-103](https://bugs.swift.org/browse/SR-103)).

Reviewer: There's far more than just one caveat. I think it's important to be realistic about these things:

```suggestion
When class inheritance and protocol conformances are used together, subclasses inherit protocol conformances from base classes, introducing additional complexity. For example, the default implementation of a protocol requirement not overridden in the conforming base class also cannot be overridden in any subclass ([SR-103](https://bugs.swift.org/browse/SR-103)).
```

> To learn more about defining and adopting protocols as well as inheritance and subclassing, see the [Protocols](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html) section and the [Inheritance](https://docs.swift.org/swift-book/LanguageGuide/Inheritance.html) section in _The Swift Programming Language_.

Reviewer: There is no barrier to being more precise here. Also, what is the distinction you mean to make between "subclassing" and "inheritance"?--If there is no distinction, use consistent terminology; otherwise readers are left to wonder if there's something they don't understand.

```suggestion
To learn more about defining and adopting protocols, see the [Protocols](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html) section in _The Swift Programming Language_. To learn more about class inheritance, see the [Inheritance](https://docs.swift.org/swift-book/LanguageGuide/Inheritance.html) section in _The Swift Programming Language_.
```

Author: “Subclassing" and "inheritance" have been used throughout this note interchangeably. For consistency and avoiding confusion, should we just use inheritance in place of subclass for the entire note?

Reviewer: Sure. Obviously, of course, you can’t change the noun “subclass.”

# Multiple Inheritance of Classes

In some programming languages, a class can inherit the interface of multiple base classes. Known as multiple inheritance, this feature can add significant complexity to the language and is unsupported in Swift. Instead, Swift allows composition of interfaces using protocols.

> Consider the following example with a protocol `Utensil` and various conforming types of specific utensils:

Reviewer: ```suggestion
Consider the following example:
```

There is no need to reiterate in words what you've written in code.

```swift
protocol Utensil { 

>     var name: String {get set}

Reviewer: ```suggestion
    var name: String { get }
```

} 

protocol ServingUtensil: Utensil {
    func serve()
} 

extension ServingUtensil {

>     func serve() {
>         print("Serving food to guests...")
>     }

Reviewer: ```suggestion
    func serve() { /* ... */ }
```

I don't think anything is served here by the example implementation.

}

protocol Fork: Utensil {
    func spear()
}

extension Fork {

>     func spear() { 
>         print("Spearing food...")
>     }

Reviewer: ```suggestion
    func spear() { /* ... */ }
```

}

struct CarvingFork: ServingUtensil, Fork {
    var name = "Carving Fork"

> }

Reviewer: A big part of the didactic value of my suggested example was that `CarvingFork`, `Ladle`, and `Spork` show how protocols can be composed in various intersecting ways. It illustrates the power of protocol composition in (IMO) a motivating way. When you use only a single concrete type, I think you also lose that key message.

This is why I'm encouraging you to trim everything from the example that doesn't serve a direct didactic purpose (like the `print` statements), to make room for the parts of the example that are important for teaching.

# Multiple Inheritance of Classes

In some programming languages, a class can have multiple superclasses. This feature is called multiple inheritance and is unsupported in Swift. However, using protocols, one can achieve a similar effect as multiple inheritance.

> For example, the protocol `Animal` and `Flyable` defines blueprints of methods like `move()`. The class `Bird` now provide concrete implementation to its conforming protocols.
> 
> ```swift
> protocol Flyable {
>     func fly()
> }
> 
> protocol Animal {
>     func move()
> }
> 
> class Bird: Animal, Flyable {
>     func move() {
>         print("Moving...")
>     }
>     
>     func fly() {
>         print("Flying...")
>     }
> }
> 
> ```
> A common pattern is when certain functionalities and characteristics are declared in a base class, and a subclass inherit those functionalities and characteristics for free. Swift provide a powerful way for protocols to implement its own requirements through an extension, and these implementations are available to conforming types by default.
> 
> Notice how there is no compiler error thrown when class `Bird` does not provide any implementation for `Animal`'s `move()` method? The extension of `Animal` has already provided a default implementation of `move()` that is available for all conforming types.
>   
> ```swift
> protocol Animal {
>     func move()
> }
> 
> extension Animal {
>     func move() {
>         print("Moving...")
>     }
> }
> ```
> 
> Conforming types can also provide their own implementations, and these declarations do not need to contain the `override` keyword. When the conforming type is a `class`, note that subclasses of a base class adopting a protocol with default implementations cannot override those declarations without the parent overriding the default implementation. 

Reviewer 2: ````suggestion
Consider the following example with a class `Bird` conforming to two protocols `Animal` and `Flyable`.

```swift
protocol Flyable {
    func fly()
}

extension Animal {
    func move() {
        print("Moving...")
    }
}

protocol Animal {
    func move()
}

extension Flyable {
    func fly() {
        print("Flying...")
    }
}

class Bird: Animal, Flyable {
    // move() and func fly() use default implementations
}
```

Since `Animal` and `Flyable` provide default implementations for their protocol requirements, `Bird` can conform to both without implementing `fly()` or `move()` and the default implementation will be used. This effect is similar to class inheritance where a subclass inherits properties, subscripts and methods from a superclass. The default implementation can be overriden by defining `fly()` or `move()` without an `override` keyword.

There is one caveat here on the interaction with subclassing. If a superclass conforms a protocol but does not override the default implementation of a protocol requirement, its subclasses cannot override the default implementation ([SR-103](https://bugs.swift.org/browse/SR-103)).
````

Reviewer 2: Feel free to incorporate parts of the suggestion above as you see fit. The main things I would like to emphasize here are:

1. Explanatory text should refrain from trying to "sell" the language. Using phrases like "powerful" and similar isn't very helpful. The reader is already using Swift, and they're trying to figure out how to solve a particular problem.
2. We should cut down on generally anything that is not necessary. Since the main example here is around default implementations, we can reasonably assume that the reader already knows how conformances work normally, so we can get away with one code example instead of two.
3. The part about "When the conforming type is a `class`, note that subclasses of a base class" -- this is actually a gotcha'/road-block, that many people hit (51 people are watching the linked JIRA), so it's better to split it out into its own paragraph instead of continuing in the same paragraph.
4. Linking the JIRA is helpful so that people can easily look things up in case they're curious about why the limitation exists and what has been discussed so far.

Author: I applied all your suggestions and agree with the above 3 main points you mentioned. Please feel free to raise more issues and suggestions so I can make sure this note can be improved to its best state before we move forward.
